pennylane/devices/qubit_mixed/sampling.py

```diff
@@ -154,95 +154,32 @@ def _measure_classical_shadow(
         state (np.ndarray): A (2^N, 2^N) density matrix for N qubits
         wire_order (qml.wires.Wires): The global wire ordering
         shots (int): Number of classical-shadow snapshots
         rng (None or int or Generator): Random seed for measurement bits
-        is_state_batched (bool): Whether the state is batched or not
 
     Returns:
         np.ndarray[int]: shape (2, shots, num_shadow_qubits).
             First row: measurement outcomes (0 or 1).
             Second row: Pauli basis recipe (0=X, 1=Y, 2=Z).
     """
-    if isinstance(mp, ShadowExpvalMP):
-        classical_shadow = ClassicalShadowMP(wires=mp.wires, seed=mp.seed)
-        bits, recipes = process_state_with_shots(
-            classical_shadow,
-            state,
-            wire_order,
-            shots,
-            rng=rng,
-        )
-        shadow = qml.shadows.ClassicalShadow(bits, recipes, wire_map=mp.wires.tolist())
-        return shadow.expval(mp.H, mp.k)
-    wire_map = {w: i for i, w in enumerate(wire_order)}
-    wires = mp.wires
-    mapped_wires = [wire_map[w] for w in wires]
-    n_snapshots = shots
-
-    # slow implementation but works for all devices
-    n_qubits = len(wires)
-
-    # seed the random measurement generation so that recipes
-    # are the same for different executions with the same seed
-    seed = mp.seed
-    recipe_rng = math.random.RandomState(seed)
-    recipes = recipe_rng.randint(0, 3, size=(n_snapshots, n_qubits))
-
-    outcomes = math.zeros((n_snapshots, n_qubits))
-    # Single-qubit diagonalizing ops for X, Y, Z
-    diag_list = [
-        qml.Hadamard.compute_matrix(),  # X
-        qml.Hadamard.compute_matrix() @ qml.RZ.compute_matrix(-np.pi / 2),  # Y
-        qml.Identity.compute_matrix(),  # Z
-    ]
-    bit_rng = math.random.default_rng(rng)
-
-    for t in range(n_snapshots):
-        for q_idx, q_wire in enumerate(mapped_wires):
-            # (A) partial trace out all other qubits to get 2x2 block for qubit q_wire
-            rho_matrix = _reshape_state_as_matrix(
-                state, _get_num_wires(state, is_state_batched=is_state_batched)
-            )
-            rho_q = math.reduce_dm(rho_matrix, [q_wire])
-
-            # (B) rotate that 2x2 block to Z-basis if recipe is X or Y
-            recipe = recipes[t, q_idx]
-            U = diag_list[recipe]
-            U = math.convert_like(U, rho_q)
-            U_dag = math.conjugate(math.transpose(U))
-            rotated = math.dot(U, math.dot(rho_q, U_dag))
-
-            # (C) probability of outcome 0 => rotated[0,0].real
-            p0 = math.clip(math.real(rotated[0, 0]), 0.0, 1.0)
-            if bit_rng.random() < p0:
-                outcomes[t, q_idx] = 0
-            else:
-                outcomes[t, q_idx] = 1
-
-    res = math.stack([outcomes, recipes]).astype(np.int8)
-    return res
+    return mp.process_density_matrix_with_shots(
+        state,
+        wire_order,
+        shots,
+        rng=rng,
+    )
```

Comment:
@astralcai @dwierichs Note that I also modified the workflow of `default.mixed` s.t. it directly uses our new routine now. Or some tests along with the `qml.shadow_expval` will still use code here same as the legacy code to be super slow.

Reply:
Sounds good! What exactly was going on with the `is_state_batched` argument before? :sweat_smile: Was it just always `False`? :thinking: